### PR TITLE
CONTRIBUTING.md for standard library

### PR DESCRIPTION
**Pull request by bobhy**
# Description

per #8810, a CONTRIBUTING.md for working on Nushell standard library.

_(Description of your pull request goes here. **Provide examples and/or screenshots** if your changes affect the user experience.)_

# User-Facing Changes

Just a text file in the source tree.  Nothing forces the developer to notice it, let alone abide by it.

# Tests + Formatting

nil

> **Note**
> from `nushell` you can also use the `toolkit` as follows
> ```bash
> use toolkit.nu  # or use an `env_change` hook to activate it automatically
> toolkit check pr
> ```

# After Submitting

nil

### CODE REVIEW

**Comment by bobhy** [comment]
The first draft has lots of placeholders like `[[text]]`.  These will be removed as the questions are resolved.
The draft will certainly need paths to be updated once #8815 lands.

**Comment by amtoine** [comment]
maybe we should just mark this as a draft and wait for #8815 to land? :yum: 
so that we'd be sure to avoid any miscomprehension and unintended merge :relieved:

**Comment by amtoine** [comment]
#8815 has been merged :partying_face: 

@bobhy, could you complete the `CONTRIBUTING.md` guidelines and ping us for a little review? :yum:

**Comment by amtoine** [comment]
i also think some parts of the `README.md` could be moved to `CONTRIBUTING.md`, e.g. the ones about implementing commands and testing :wink:

**Comment by bobhy** [comment]
Ready for review, woo-hoo!

File: crates/nu-std/README.md
Comment (amtoine) [inline crates/nu-std/README.md:74]: with the current module system, that would not work :eyes: 
```bash
> use std [log assert]
Error: nu::parser::export_not_found

  × Export not found.
   ╭─[entry #2:1:1]
 1 │ use std [log assert]
   ·          ─┬─
   ·           ╰── could not find imports
   ╰────
```

however, applying the following would fix it
```diff
-use std log assert
+use std ['log debug' assert]
```

File: crates/nu-std/README.md
Comment (amtoine) [inline crates/nu-std/README.md:80]: maybe a period here for consistency with the rest of the document? :yum:

File: crates/nu-std/README.md
Comment (amtoine) [inline crates/nu-std/README.md:12]: i think this will be tideous to maintain and to scale...

for the time being, i'd leave the command which is great but remove the big table :+1:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:37]: i think that should be `[below](#the-pr)` to jump to the anchor.

otherwise, when i read the document on your PR branch, it wants to bring me to the invalid https://github.com/bobhy/nushell/blob/stdlib_contributing/crates/nu-std/the_pr :scream:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:44]: i you mean "solve the review threads when you, as the author of the PR, thinks you have addressed them", i'd strongly discourage that :yum: 

as a reviewer, i find this super confusing and tideous when someone else than me, either the author or another reviewer, closes a thread i've opened for me :confused: 
i've opened the thread, so i'd like to know if the changes have been properly addressed :smirk: 
when a thread is closed by someone else, it does not appear anymore in the review feed and the reviewer would have to scroll all the resolved threads to find what has been answered, if the changes have been approved, etc, etc...

of course, that is my experience and i find that very confusing myself, but i'd really like not mentionning that opportunity :eyes: 

=> i also think specifying in the thread what commit(s) should addressed the issue is very sweet for reviewers :relieved:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:78]: booooh we love `vim` :stuck_out_tongue_closed_eyes:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:95]: even better => toggle the "create draft pull request" with the arrow next to the green default "create pull request" OR click on the "still a work in progress? convert to draft" button on the status side pannel!

that would mark the PR gray as a "true" DRAFT :wink:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:110]: just wondering if we should only write `crates/nu-std/...` instead of the generic `/path/to/...` ?

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:127]: almost same remark here => that might be made somewhat smaller by directly linking to the two modules without code blocks :+1:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:130]: the file is called `mod.nu` :wink:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:135]: this is actually incorrect but it's a recent change in #8815 :relieved: 

should be
```
   export use foo *    # command doesn't update environment
   export-env {
          use bar *    # command *does* update environment
   }
```

:bulb: modules are automatically loaded from `crates/nu-std/lib/` as `.nu` files thanks to the library loader in `crates/nu-std/src/lib.rs` :+1:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:165]: here it should be `cargo run --` and not plain `nu` :yum: 

as the library is embedded in the `nu` binary, they should be in sync together => the easiest to do that is
- to use `cargo run -- ...`
- to use the `toolkit` module in the root of the project :wink:

**Review by amtoine — CHANGES_REQUESTED** [review]
thanks a looooot for this, that's amazing :sparkles: 

mainly little tweaks, a few more important ones :wink: 

:exclamation: the document is very complete, i like that, but it's quite long already => i think a little table of content with `[<name>](#anchor-name)` links at the top would be great to navigate this quickly :yum:

File: crates/nu-std/README.md
Comment (bobhy) [reply crates/nu-std/README.md:74]: I definitely should have verified the example works as documented.  I'll update this.
But why does `use` require second argument to be an array? Doesn't `use` take any number of arguments (rather than exactly 2, the second of which is an array ...)?

File: crates/nu-std/README.md
Comment (bobhy) [reply crates/nu-std/README.md:80]: yes.

File: crates/nu-std/README.md
Comment (bobhy) [reply crates/nu-std/README.md:12]: I'll telescope the table, but want to show some sample results.  If user wants to see the whole list in the current version, s/he can run the command then.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:37]: yes.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:44]: My use of "resolve" is confusing, maybe "address" is closer to the mark?  I mean that the owner of the PR should change the code or convince the reviewer it's fine as is.
Also, is it the norm on this project that the reviewer who opens the thread can be relied on to close it when satisfied?  Github allows other participants to "resolve" the thread, and I've seen projects where the owner of the PR is the only one who's motivated to clean up the threads.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:78]: I knew it!  (where are the emacs bigots?) 
I'll fuzzy this up...

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:95]: "draft" PR, gotcha.  Don't want to depend on the current github UI.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:110]: I have to be consistent about paths, one way or the other....

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:127]: I wanted to highlight the "note the use of ..." outside the code block.  I'll come up with some other presentation for this.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:130]: Right.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:135]: OK, the advice is "monkey see, monkey do", but it means std is becoming less and less like a typical user-written module library.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:165]: Let me understand better.  Is it the case that the code for std is not loaded from the `.nu` files any longer?
How then can the developer try making changes to std.nu and friends? Does s/he have to rebuild the interpreter after making a script change?
Is there any dependency on an environment variable to find any of the std `.nu` files?
Does `$env.NU_LIB_DIRS` still function to allow you to omit the full path in a `use <module>` for modules other than `std`.  How does Nushell tell the difference?

I want to assume a std developer already has some familiarity with modules in general, so I think I need to be careful to point out where they're different so s/he's not misled by too much knowlege.

In any case, I think the point of your correction is that developer must be sure to run the instance of the interpreter that's in the current source tree (thus `cargo run`, and not run the interpreter that's somewhere in his/her $env.PATH, which might be a different version.  Is that it?

File: crates/nu-std/README.md
Comment (amtoine) [reply crates/nu-std/README.md:74]: > I definitely should have verified the example works as documented. I'll update this.
>
no worries, this is why humanity invented the Review :wink: 

> But why does `use` require second argument to be an array? Doesn't `use` take any number of arguments (rather than exactly 2, the second of which is an array ...)?
>
i do not know the "why" but...
```bash
>_ use std 'log debug' assert
Error: nu::parser::extra_positional

  × Extra positional argument.
   ╭─[entry #1:1:1]
 1 │ use std 'log debug' assert
   ·                     ───┬──
   ·                        ╰── extra positional argument
   ╰────
  help: Usage: use <module> (members)
```
according to `help use`, it takes a `members` but not a `...members` :open_mouth:

File: crates/nu-std/README.md
Comment (amtoine) [reply crates/nu-std/README.md:12]: that sounds a great compromise :+1:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:78]: i've not seen Emacs pop up in the `nushell` community for now :smirk:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:95]: for this i'm not sure you can abstract that much away from the UI unfortunately :thinking: 
we could always use `gh` but that's probably too advanced for this document :relieved:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:110]: i might prefere `crates/nu-std/...` a bit
- removes a bit of noise from `/path/to/...`
- a bit shorter

BUT it assumes you are in the root of `nushell` (which is not that much of a strong assumption?)

:bulb: i do not have a strong preference here

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:135]: ehe nope it's not anymore :yum: 

we decided, for now at least, to embed the library in the `nu` binary directly => the direct consequence is that we can not rely on paths like `crates/nu-std/lib/<mod>.nu` because the user won't have that file at all on the system!
only specifying `<mod>` is all the automatic loader (in `crates/nu-std/src/lib.rs`) needs to know :+1:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:165]: > Let me understand better.
>
:+1: 

> Is it the case that the code for std is not loaded from the `.nu` files any longer?
>
it is not.
as i said in the thread just above, the `nu-std` code loads the modules by name just before the REPL starts or a script runs.

> How then can the developer try making changes to std.nu and friends? Does s/he have to rebuild the interpreter after making a script change?
>
yes they should rebuild => hence the `cargo run`
you could have an old version of `nu` installed which could make `nu .../tests/run.nu` behave unexpectedly.
PLUS we can not rely, in the end, on the user having the files on their filesystem, because *apriori* they won't!

> Is there any dependency on an environment variable to find any of the std `.nu` files? Does `$env.NU_LIB_DIRS` still function to allow you to omit the full path in a `use <module>` for modules other than `std`. How does Nushell tell the difference?
> 
the `nu-std` loader loads the whole library before startup and make it available to the user as `std` => you just have to call `use std` to import all the library.
> **Note**
> it's a bit like the library "being in `$env.NU_LIB_DIRS`", but it's really at the binary level => `nushell` does not look up in `$env.NU_LIB_DIRS` for `std`, it just loads it directly from the binary.

> I want to assume a std developer already has some familiarity with modules in general, so I think I need to be careful to point out where they're different so s/he's not misled by too much knowlege.
> 
absolutely!
the only "new" rule to follow here is to use the name of the modules only not their paths :relieved: 

> In any case, I think the point of your correction is that developer must be sure to run the instance of the interpreter that's in the current source tree (thus `cargo run`, and not run the interpreter that's somewhere in his/her $env.PATH, which might be a different version. Is that it?
>
that's exactly right :ok_hand: 

the worse really would be having a `nu` binary which does not feature the `std` loader / relies on path to the `std` modules...
that would very likely lead to unexpected behaviour or more probably just crash :scream: 
in the end, i did not try such a thing, but from what i wrote in the loader, that does not look right :yum: 

hope all of this is a bit clearer :relieved:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:44]: > My use of "resolve" is confusing, maybe "address" is closer to the mark? I mean that the owner of the PR should change the code or convince the reviewer it's fine as is. 
>
oh yeah perfect!
then i'd keep "address" for this very action :wink: 

as "resolve" is the action *GitHub* refers to on the button to "close" the threads, that might be misleading?

> Also, is it the norm on this project that the reviewer who opens the thread can be relied on to close it when satisfied?
>
i do not know if there's a norm here :thinking: 
maybe in the main `CONTRIBUTING.md`?

these are my two cents :relieved: 
i really find that super confusing not seing my threads where i left them :confused: 

> **Note**
> `nushell` does not require approvals nor thread resolution to merge a PR, so we do not even really *need* to close threads before merging.
> i see them as indications for another maintainer that wants to answer the question "can i merge this PR" => an opened thread without clear closure might indicate a non-quite-ready PR

> Github allows other participants to "resolve" the thread, and I've seen projects where the owner of the PR is the only one who's motivated to clean up the threads.
>
imo, that looks like a pretty bad practice :thinking: 

i think review threads ARE the review itself.
they are the place where maintainers and reviewers discuss with the author what needs to be changed, what could be changed.
if the author closes the threads, that looks almost like a hack where they say "i wont' work on this and you won't see this unless you scroll all your closed threads" => that could make the "thread resolution" requirement branch protection rule happy even though nothing has been addressed :open_mouth:

it's like doing a rebase / force-push without a reason => that disqualifies all the previous review and commits because i do not have any guarantee as a reviewer that the rebase did not break any thing :yum: 

in the end i'd say three things
- there is no consensus on this, this is from my personal review experience
- i find closing my threads for me really confusing and a waste of time as i need to go and get them back...
- i think the responsibility to close a thread is on the person who opened it (or maybe a very high-permission maintainer who really knows what their doing :/)

Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:44]: (not targeting your force-push here @bobhy, this was a good one :wink:)

Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:44]: i'm really just expressing very general review guidelines but i can not resist :eyes:
i think a good review workflow is really important :yum: 

TL;DR: changing "resolve" to "address" in this document would be just perfect :blush:

File: crates/nu-std/README.md
Comment (amtoine) [reply crates/nu-std/README.md]: :+1:

File: crates/nu-std/README.md
Comment (amtoine) [reply crates/nu-std/README.md]: :+1:

**Comment by fdncred** [comment]
This PR isn't an "emergency" but I do think we need to circle back at some point and finish it. The force-push kind of set us back.

File: crates/nu-std/README.md
Comment (amtoine) [inline crates/nu-std/README.md:51]:
```diff
-See [CONTRIBUTING.md](https://github.com/nushell/nushell/blob/main/crates/nu-std/CONTRIBUTING.md) for details.
+See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
```
should work.

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:162]:
```diff
-2. Use `log info` to provide a verbose progress messages that the user can optionally enable for troubleshooting. 
+2. Use `log info` to provide verbose progress messages that the user can optionally enable for troubleshooting. 
```
right?

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [inline crates/nu-std/CONTRIBUTING.md:176]: i'd apply
```diff
-  Change 'ERROR' to 'INFO' or 'DEBUG' for increasing verbousity.
+  Change 'ERROR' to `INFO` or `DEBUG` for increased verbosity.
```

**Review by amtoine — CHANGES_REQUESTED** [review]
a few typos and broken links :+1: 

> **Note**
> i think the `md` style will have to be reworked at some point (lots of trailing whitespaces and the format looks a bit odd to me) **BUT** the content of `CONTRIBUTING.md` is ok, so i'd like to land that PR and see all this in some later "cleanup" phase :relieved:

File: crates/nu-std/README.md
Comment (bobhy) [reply crates/nu-std/README.md:51]: Actually, no.  Your proposed change won't work in local repo, only online, thanks to github magic.  I think the risk that the online target will disappear in the future is low.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:79]: "token" means a trivial change, a placeholder.  The idea is open a PR and mark it as draft as soon as you get started, before you have any real changes to commit.  This, so you can draft up a "functional spec" in the commends of the PR as you proceed.

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:139]: Yep.

File: crates/nu-std/README.md
Comment (amtoine) [reply crates/nu-std/README.md:51]: oooh that is for local repos :open_mouth: :+1:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md:79]: oh i see, i just did not know that way of saying this :relieved:

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:151]: We should describe what the 2 elements are actually for.  I tried some experiments and documented 
my best guess (and added comments to `lib.rs`, which should also be reviewed.)

File: crates/nu-std/CONTRIBUTING.md
Comment (bobhy) [reply crates/nu-std/CONTRIBUTING.md:176]: Doesn't press my "bad grammar" button, but OK...

**Comment by bobhy** [comment]
> a few typos and broken links +1
> 
> > Note
> > i think the `md` style will have to be reworked at some point (lots of trailing whitespaces and the format looks a bit odd to me) **BUT** the content of `CONTRIBUTING.md` is ok, so i'd like to land that PR and see all this in some later "cleanup" phase relieved

Don't clean the trailing whitespace up too aggressively!  I use 2 blanks at the end of the line to force MD to start a new rendered line;  I use that a lot in lists.
I did some (nonbreaking) line breaks in the souirce to keep them short.

File: crates/nu-std/README.md
Comment (bobhy) [reply crates/nu-std/README.md:51]: Hmm, do we want the link to open, or do we want the path to be obvious?

**Comment by amtoine** [comment]
> Don't clean the trailing whitespace up too aggressively! I use 2 blanks at the end of the line to force MD to start a new rendered line; I use that a lot in lists. I did some (nonbreaking) line breaks in the souirce to keep them short.
>
yep, of course, do not worry :relieved: 
that's easy to filter out :wink:

File: crates/nu-std/README.md
Comment (amtoine) [reply crates/nu-std/README.md:51]: the absolute HTTP path is good :+1:

Comment (amtoine) [reply crates/nu-std/README.md:51]: :heavy_check_mark:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md]: :heavy_check_mark:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md]: :heavy_check_mark:

File: crates/nu-std/CONTRIBUTING.md
Comment (amtoine) [reply crates/nu-std/CONTRIBUTING.md]: :heavy_check_mark:

**Comment by bobhy** [comment]
I innocently pulled latest changes from main and rebased my 3 file changes on top of that int order to keep the PR fresh and mergable.  But now I see that the PR shows 35 files modified.  The changes seem to be recent merges into main, but I've completely lost confidence in the integrity of this PR!  I'm going to close it and open a new one with just my intended changes; let's see whether the new one can get through review a bit more expeditiously...

**Comment by bobhy** [comment]
This PR has gone off the rails and has been deprecated in favor of #8894, q.v.

**Pull request closed without merging**
